### DISCUSSION TITLE
How to handle numerical errors in pmf probabilities summing to 1?

### DISCUSSION THREAD

**Discussion by elray1** [hubverse-org/hubEvals#64] [General]
For hubEvals, we're using the scoringutils package, which in turn uses scoringRules to calculate the ranked probability score (rps).  Between hubValidations, scoringutils, and scoringRules, there are three different checks that are being done for whether or not class probabilities sum to 1:

- hubverse validations [require that class probabilities sum to 1 within a tolerance of +/- the square root of machine epsilon](https://github.com/hubverse-org/hubValidations/blob/4f511c2eeb371a194ad9583e1007fa568d1022dc/R/check_tbl_value_col_sum1.R#L52) (this is the default tolerance in `all.equal` in R), which depends on machine architecture, but for me is about 1.5e-8.
- The scoringutils package [requires that class probabilities sum to 1 within a tolerance of +/- 1e-10](https://github.com/epiforecasts/scoringutils/blob/e73e2b872b66d59fa1244958b78d55fab399f31e/R/metrics-nominal.R#L73-L80).
- The scoringRules package [requires that the sum of the class probabilities minus 1 is <= machine epsilon](https://github.com/FK83/scoringRules/blob/6a246753c5f76dada6e33c048f26f4e475733c84/R/rps.R#L59), which is about 2.2e-16 for me.

There are two issues here:

1. I'm pretty sure the scoringRules validations are more stringent than we can guarantee we'll be able to satisfy.  Even normalizing a vector by its sum could yield something that fails this condition:
```
> set.seed(42)
> x <- abs(rnorm(n = 100, mean = 0.5))
> x <- x / sum(x)
> isTRUE(abs(sum(x) - 1) > .Machine$double.eps)
[1] TRUE
> abs(sum(x) - 1)
[1] 4.440892e-16
> .Machine$double.eps
[1] 2.220446e-16
```
So I think we should file an issue about this at scoringRules, but I'm not knowledgeable enough about this to be confident in what to suggest that they change their check to.  Should it just be `sqrt(.Machine$double.eps)`?

2. We may end up in a situation where a pmf submission passed hubverse validations, but fails to pass scoringutils and/or scoringRules validations.  How should we handle this?

Three ideas:
1. Allow `scoringutils` and/or `scoringRules` to throw errors if their more-stringent criteria are not met.
2. Try to call the `scoringrules` function, but catch errors.  If errors are thrown, normalize the class probabilities on behalf of the user and issue a warning instead of an error.
3. Always normalize class probabilities. We're assuming the model outputs this function is being handed passed validations, so the class probabilities come close to 1 and we're doing some minor cleanup on behalf of the package user.

**Comment by nickreich** [comment]
Regarding options 2 and 3 above for the "ideas" section: It wasn't clear to me if normalizing would solve things, given the example you provide above. It seems like the simplest thing for us to do at hubverse would be idea 1 (just return the errors). How much would that be then putting back on users to correct/normalize things? And is that work feasible. In general, it seems like idea 1 is the simpler and more ideal solution (less for hubverse to maintain down the road).

**Comment by zkamvar** [comment]
It seems like it's worth bringing this up with the scoringRules maintainers, especially since you cannot guarantee that a normalized vector will sum to one. From what I understand the `sqrt(.Machine$double.eps)` dates back to numerical recipes (https://scicomp.stackexchange.com/a/14356).

**Comment by annakrystalli** [comment]
First a question. Have we established that submissions files (i.e. outputs of most models) are more likely than not to fail the `scoringRules` criteria? Searching the, admittedly limited, number of issues in their repo, I can't see anyone else having reported a problem with their tolerance. So it could be that we could just change our tolerance in `hubValidations` with minimum disruption.

Having said that, I feel the standard R equality test tolerance should be an acceptable tolerance for R stats packages. As such I also vote for contacting the `scoringRules` maintainers. Instead of requiring them to change the tolerance, however, perhaps a better suggestion would be for them to introduce a `tolerance` argument (like the one `all.equal()` has) that they can set to whatever tolerance they prefer as default but we can use to override.

**Comment by elray1** [comment]
I'm finally getting back to this.  Thanks all for your comments above!  Here are some thoughts:

I don't know how pervasive this is in general, but the main place I want to do score computations with pmf outputs is related to FluSight, and the more restrictive sum-to-1 constraints are not satisfied there.  Here's a semi-reproducible example (my working directory is the root of a local clone of [the Flusight hub](https://github.com/cdcepi/FluSight-forecast-hub)):
```
hub_con <- hubData::connect_hub(".")
pmf_preds <- hub_con |>
  dplyr::filter(output_type == "pmf", target == "wk flu hosp rate change") |>
  dplyr::collect()

pmf_preds |>
  dplyr::group_by(reference_date, target, horizon, target_end_date, location, output_type, model_id) |>
  dplyr::summarize(
    sum_value = sum(value),
    abs_diff_from_1 = abs(sum_value - 1)
  ) |>
  dplyr::arrange(desc(abs_diff_from_1)) |>
  head(10)
```

This has the following output:
```
# A tibble: 10 × 9
# Groups:   reference_date, target, horizon, target_end_date, location, output_type [10]
   reference_date target                  horizon target_end_date location output_type model_id           sum_value abs_diff_from_1
   <date>         <chr>                     <int> <date>          <chr>    <chr>       <chr>                  <dbl>           <dbl>
 1 2024-12-21     wk flu hosp rate change       2 2025-01-04      51       pmf         FluSight-ens_q_cat      1.00        2.84e- 8
 2 2024-12-21     wk flu hosp rate change       1 2024-12-28      40       pmf         FluSight-ens_q_cat      1.00        1.74e- 9
 3 2024-12-21     wk flu hosp rate change       1 2024-12-28      08       pmf         FluSight-ens_q_cat      1.00        3.14e-10
 4 2023-12-23     wk flu hosp rate change       1 2023-12-30      51       pmf         FluSight-ens_q_cat      1.00        4.90e-11
 5 2023-12-23     wk flu hosp rate change       1 2023-12-30      27       pmf         FluSight-ens_q_cat      1.00        4.79e-11
 6 2024-01-06     wk flu hosp rate change       2 2024-01-20      54       pmf         FluSight-ens_q_cat      1.00        4.76e-11
 7 2024-01-27     wk flu hosp rate change       1 2024-02-03      42       pmf         FluSight-ens_q_cat      1.00        4.52e-11
 8 2023-12-30     wk flu hosp rate change       1 2024-01-06      19       pmf         FluSight-ens_q_cat      1.00        4.48e-11
 9 2024-02-03     wk flu hosp rate change       1 2024-02-10      12       pmf         FluSight-ens_q_cat      1.00        4.37e-11
10 2024-01-13     wk flu hosp rate change       1 2024-01-20      17       pmf         FluSight-ens_q_cat      1.00        4.22e-11
```

The largest of those, `2.84e- 8`, is I think actually larger than the square root of machine epsilon on my machine?
```
> sqrt(.Machine$double.eps) < 2.84e-8
[1] TRUE
```

The largest 3 values are all bigger than the cut off of 1e-10 used by `scoringutils`, and all of these are larger than the cut off used by `scoringRules`.

If I try to compute scores with these, I get errors and weird results.

The reason I don't like option 1 on my list above (throw errors) is that with this situation, we really can't evaluate these predictions. Note that one set of users of hubEvals is us! The evals dashboard would not be able to support evaluations of pmf predictions.

So, I'm in favor of Anna's suggestion.

**Comment by elray1** [comment]
I filed an issue with the to-dos that came out of this here: https://github.com/hubverse-org/hubEvals/issues/74
